### PR TITLE
fs_notify generates events on wrong directory paths

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
## Proposed changes

This PR fixes the issue related to raising the events for the wrong `directory` when  `watching multiple directories`.
Uses the `inotify` fix that helps to extract the exact path on which the event has been raised.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
https://github.com/thin-edge/thin-edge.io/issues/1453

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (initard) [inline crates/common/tedge_utils/src/fs_notify.rs:72]: It would be best if only  one of these data structures existed, not both. The `description` data structure's goal is to describe a watched file or directory. 

It's composed of 
> {
>  "parent_path": {Option<"file_name">: [FileEvent]},
> } 

This was done to distinguish between watching a directory or watching a file, such that for watching a file it looks like:

> {
>  "parent_path": {**Some(**"file_name"**)**: [FileEvent]},
> } 

And watching a directory it looks like:

> {
> "parent_path": {**None**: [FileEvent]},
>} 


It was needed because the event id was not there. Now that we have this id, i think the data structure can be changed altogether and we can use the new one proposed and find a way of adding the vector (or set) of FileEvent(s).

I leave it up to you to decide the new data structure but here is an idea. 

```python 
{
    "c_int": ("full_path", [FileEvent_1, FileEvent_2]),
}
``` 
I think a data structure like this fully describes an inotify event and removes a lot of extra code lower down.

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (didier-wenzek) [inline crates/common/tedge_utils/src/fs_notify.rs:290]: This 3 levels loop is a sign that the data structure `watchers.description` is not adapted for the task. We are desperately looking for the file matching the event.  Now that a watch descriptor id is accessible from the inotify event, we can simply use a map that gives all what we need given such a watch descriptor id.

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (didier-wenzek) [inline crates/common/tedge_utils/src/fs_notify.rs:317]: I don't understand this comment.

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (didier-wenzek) [reply crates/common/tedge_utils/src/fs_notify.rs:72]: I agree with @initard, one should have single structure that leverages the watch descriptor id that has been recently added by the `inotify` crate.

```
struct WatchDescriptor {
    description: HashMap<c_int, (PathBuf, Vec<FileEvent>)>,
}
```

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
The recent updates of the `inotify` crate gives an opportunity to have a code that is simpler, clearer and more efficient. Avoiding this refactoring will only add confusion.

File: plugins/c8y_configuration_plugin/src/main.rs
Comment (initard) [inline plugins/c8y_configuration_plugin/src/main.rs:325]: is this needed?

Comment (PradeepKiruvale) [reply plugins/c8y_configuration_plugin/src/main.rs:325]: Yes, to add a watcher to any directory, that directory should exists.

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (PradeepKiruvale) [reply crates/common/tedge_utils/src/fs_notify.rs]: The above-mentioned structure does not fit here. So, using the below structure.

struct WatchDescriptor {
    description: HashMap<c_int, Vec<FileOrDir>>,
}
pub struct FileOrDir {
    pub dir_path: PathBuf,
    pub file_name: Option<String>,
    pub masks: Vec<FileEvent>,
}

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (initard) [inline crates/common/tedge_utils/src/fs_notify.rs:94]: `get_mask_set_for_directory` the name and what the function is doing does not seem to be right

Comment (PradeepKiruvale) [reply crates/common/tedge_utils/src/fs_notify.rs:94]: That's what exactly it does. It just collects all the `event masks` to that specific `directory`.

Comment (initard) [reply crates/common/tedge_utils/src/fs_notify.rs:94]: It's more "for descriptor id" or "for root/parent directory"

Comment (initard) [reply crates/common/tedge_utils/src/fs_notify.rs:94]: because I would expect to give a directory as input to something called get_mask_set_for_directory, but its actually taking an id as input

Comment (PradeepKiruvale) [reply crates/common/tedge_utils/src/fs_notify.rs:94]: Yes, it's for the `wid`, I missed that part, I will rename that.

File: plugins/c8y_configuration_plugin/src/main.rs
Comment (initard) [reply plugins/c8y_configuration_plugin/src/main.rs:325]: I'm just confused as to why a change was needed in the c8y configuration plugin? :smile:

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
This code seems to have wrong assumptions about [`Inotify.add_watch()`](https://docs.rs/inotify/latest/inotify/struct.Inotify.html#method.add_watch) and on the conditions when a new watch event is created.

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (initard) [reply crates/common/tedge_utils/src/fs_notify.rs:87]: Yes, you are right that `HashMap<c_int, EventDescription>` should be what we need here however sadly this is not possible because if a file does not exist you can not watch it. There is no `c_int` for this event.

Because of this issue, internally all watchers are directory watchers. 

This means that `c_int` can map to one or more events, hence the use of vector. When an inotify event is triggered, this event is matched on two things:
1. the `c_int`
2. the [name](https://docs.rs/inotify/latest/inotify/struct.Event.html#structfield.name) of the file the event was triggered for (which is an Option<file_name>)

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (initard) [reply crates/common/tedge_utils/src/fs_notify.rs:227]: This is an update from the inotify crate. `watches().add()` is the correct way of doing it now. `add_watch()` is deprecated.

**Comment by PradeepKiruvale** [comment]
> This code seems to have wrong assumptions about [`Inotify.add_watch()`](https://docs.rs/inotify/latest/inotify/struct.Inotify.html#method.add_watch) and on the conditions when a new watch event is created.
This is how the add_watch() works.
1. The watcher can only be added to the existing paths. It could be a file or a directory.
2. If all the files in a directory have to be watched for similar events then we can maintain only one `event description` entry per `wid` (watch description id).
    If each file must be monitored for various different events, we must maintain different `event description` entries.  Still, the `wid` will remain the same. So, we have a vector of `event descriptions` for a given `wid`.

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (PradeepKiruvale) [reply crates/common/tedge_utils/src/fs_notify.rs:105]: This is useful if `all/multiple files` have to be watched for `create` and other modification `events`. So, in this case the path to watch will be the `directory` path not the `file path`. So, there will be only one `description id` for all the files under watch.

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (PradeepKiruvale) [reply crates/common/tedge_utils/src/fs_notify.rs:272]: Here the `cross verification` of the raised event is done to check if the event raised contains the `file name` and the `event masks` that were supposed to be watched for.

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (PradeepKiruvale) [reply crates/common/tedge_utils/src/fs_notify.rs]: It was kind of the wrong function signature. I refactored it's not applicable anymore.

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (PradeepKiruvale) [reply crates/common/tedge_utils/src/fs_notify.rs:84]: Yes, there will be one `event id` for watching different files in the same `directory`.
For example in a directory `/etc/tedge/operations/c8y` if we want to watch `c8y_ops` & `c8y_ops2` (Assuming the files were not existing earlier) then the watcher will be added to the directory. When some event triggers on this directory, then we need to check on which files these events were raised. If the event is raised for the files we are interested in then we can capture those events and forward them for further processing.

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (didier-wenzek) [inline crates/common/tedge_utils/src/fs_notify.rs:264]: I have many comments on this return type.

* The empty string `""` is used to mean nothing. => The return type should wrap an `Option<String>` instead of a `<String>`.
* I also wonder how this function can fail. It happens that this is only when the event is unexpected. i.e no path matches the event. => Return `None` when there is no match and the `Result` type is then useless.
* What you need at the end is a `PathBuf` not a `String`. => Return a `PathBuf`.

```suggestion
    ) -> Option<PathBuf> {
```

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (didier-wenzek) [inline crates/common/tedge_utils/src/fs_notify.rs:265]: This comment must say *why* the event always has a file name.

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (didier-wenzek) [reply crates/common/tedge_utils/src/fs_notify.rs:272]: Using a `Vec<EventDescription>` is not the appropriate struct then.

What you need is a `HashMap<<Option<String>, EventDescription>` since you are always looking for *the* event that matches either some file name (the name of the file where something happened) or none (meaning any file in that directory).

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (didier-wenzek) [reply crates/common/tedge_utils/src/fs_notify.rs:84]: Okay it's clearer now. However, this deserves a comment describing the structure and the why.

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
There are still some points to improve.

The structure of this crate can be improved in a following PR:
* Why some functions are related to a struct as `NotifyStream::get_full_path_and_mask` and some others not as `pipe_masks_into_watch_mask`.
* The `WatchDescriptor` struct seems a bit unnecessary, as the map of events can be a field of `NotifyStream`.
* The `pipe_masks_into_watch_mask` is badly named. This is an `union` or an `or` not a `pipe` operation.
*

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (PradeepKiruvale) [reply crates/common/tedge_utils/src/fs_notify.rs:272]: I think the proposed data struct does not work because 

- When the same file name can be part of two directories.
- Also when it's `None`, then which one to pick up?

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (PradeepKiruvale) [reply crates/common/tedge_utils/src/fs_notify.rs:264]: Thanks, changed it to `Option<PathBuf>`

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (PradeepKiruvale) [reply crates/common/tedge_utils/src/fs_notify.rs:89]: added a comment

File: crates/common/tedge_utils/src/fs_notify.rs
Comment (didier-wenzek) [reply crates/common/tedge_utils/src/fs_notify.rs:278]: I think you missed my point.

I propose to use a `HashMap<Option<String>, EventDescription>` for *each* directory under watch.

```
description: HashMap<c_int, HashMap<Option<String>, EventDescription>>,
```

Look the current code for `get_full_path_and_mask` that use a `Vec<EventDescription>`.
* This code iterates over the event descriptions to find *the* description, if any, that match the current event file. Using a map from file names to event descriptions, this can be done directly. 
* The current code also iterates over the event descriptions to find *the* event description attached to the directory, if any, i.e the entry with *no* filename. This is for that use case that the keys of the map must be options.

**Review by didier-wenzek — APPROVED** [review]
Approved - to unblock the awaited https://github.com/thin-edge/thin-edge.io/issues/1448. However, the code can really be improved, and I expect a following refactoring PR from @PradeepKiruvale.

**Pull request merged**
